File: java/gov/nasa/gsfc/gmsec/api/field/U16Field.java
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


/**
 * @file U16Field.java
 */

package gov.nasa.gsfc.gmsec.api.field;

import gov.nasa.gsfc.gmsec.api.U16;
import gov.nasa.gsfc.gmsec.api.jni.field.JNIU16Field;


/**
 * Specialized Field class that can be used to retain an unsigned 16-bit value.
 */
public class U16Field extends Field
{
	/**
	 * This constructor is for internal GMSEC API use only.
	 * @param field A JNIU16Field object.
	 */
	public U16Field(JNIU16Field field)
	{
		setInternal(field);
	}


	/**
	 * Default constructor.
	 *
	 * @param name Name of the field.
	 * @param data Data to associate with the field.
	 *
	 * @throws IllegalArgumentException Thrown if the field name is null, or contains an empty string.
	 * @throws IllegalArgumentException Thrown if the data value is null.
	 */
	public U16Field(String name, U16 data) throws IllegalArgumentException
	{
		validateName(name);

		if (data == null)
		{
			throw new IllegalArgumentException("U16Field object is null");
		}

		setInternal(new JNIU16Field(name, data));
	}


	/**
	 * Copy constructor.
	 *
	 * @param other The other U16Field object to copy.
	 *
	 * @throws IllegalArgumentException Thrown if the given U16Field object is null.
	 */
	public U16Field(U16Field other) throws IllegalArgumentException
	{
		if (other == null)
		{
			throw new IllegalArgumentException("U16Field object is null");
		}

		setInternal(new JNIU16Field((JNIU16Field) Field.getInternal(other)));
	}


	/**
	 * Returns the data stored within the U16Field.
	 *
	 * @return A U16 object.
	 */
	public U16 getValue()
	{
		JNIU16Field jU16Field = (JNIU16Field) getInternal();

		return jU16Field.getValue();
	}
}

File: framework/src/gmsec4/c/field/u16_field.cpp
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */



/** @file u16_field.cpp
 *
 *  @brief This file contains functions for the management of an U16Field object.
 */


#include <gmsec4/c/field/u16_field.h>

#include <gmsec4/field/U16Field.h>

#include <gmsec4/Exception.h>
#include <gmsec4/Status.h>


using namespace gmsec::api;


GMSEC_Field CALL_TYPE u16FieldCreate(const char* fieldName, GMSEC_U16 value, GMSEC_Status status)
{
	U16Field* field = NULL;
	Status    result;

	try
	{
		field = new U16Field(fieldName, value);
	}
	catch (Exception& e)
	{
		result = Status(e);
	}

	if (status)
	{
		*(reinterpret_cast<Status*>(status)) = result;
	}

	return field;
}


GMSEC_U16 CALL_TYPE u16FieldGetValue(const GMSEC_Field field, GMSEC_Status status)
{
	const Field*    f        = reinterpret_cast<const Field*>(field);
	const U16Field* u16Field = dynamic_cast<const U16Field*>(f);
	GMSEC_U16       value    = 0;
	Status          result;

	if (u16Field)
	{
		value = u16Field->getValue();
	}
	else
	{
		result = Status(FIELD_ERROR, INVALID_FIELD, "Field object is not a U16Field");
	}

	if (status)
	{
		*(reinterpret_cast<Status*>(status)) = result;
	}

	return value;
}
